File: Source/NativeApi/NativeApi/Api/Sizer/GridBagSizer.cs
#pragma warning disable
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Alternet.Drawing;

namespace NativeApi.Api
{
    //https://docs.wxwidgets.org/3.2/classwx_grid_bag_sizer.html
    //A wxSizer that can lay out items in a virtual grid like a wxFlexGridSizer but in
    //this case explicit positioning of the items is allowed using wxGBPosition,
    //and items can optionally span more than one row and/or column using wxGBSpan.
    public class GridBagSizer : FlexGridSizer
    {
        // Constructor, with optional parameters to specify the gap between the rows and columns.
        public static IntPtr CreateGridBagSizer(int vgap = 0, int hgap = 0) => default;

        // Called when the managed size of the sizer is needed or when layout needs done. 
        public static Int32Size CalcMin(IntPtr handle) => default;

        // Return the sizer item located at the point given in pt, or NULL if there is
        // no item at that point.
        public static IntPtr FindItemAtPoint(IntPtr handle, Int32Point pt) => default;

        // Return the sizer item for the given grid cell, or NULL if there is no item
        // at that position.
        public static IntPtr FindItemAtPosition(IntPtr handle, Int32Point pos) => default;

        // Return the sizer item that has a matching user data(it only compares
        // pointer values) or NULL if not found.
        public static IntPtr FindItemWithData(IntPtr handle, IntPtr userData) => default;

        // Get the size of the specified cell, including hgap and vgap.
        public static Int32Size GetCellSize(IntPtr handle, int row, int col) => default;

        // Get the size used for cells in the grid with no item.
        public static Int32Size GetEmptyCellSize(IntPtr handle) => default;

        // Called when the managed size of the sizer is needed or when layout needs done.
        public static void RepositionChildren(IntPtr handle, Int32Size minSize) { }

        // Set the size used for cells in the grid with no item.
        public static void SetEmptyCellSize(IntPtr handle, Int32Size sz) { }

        // Adds the given item to the given position.
        public static IntPtr Add(IntPtr handle, IntPtr window, Int32Point pos,
            Int32Size span, int flag /*= 0*/, int border /*= 0*/,
            IntPtr userData) => default;

        // Adds the given item to the given position.
        public static IntPtr Add2(IntPtr handle, IntPtr sizer, Int32Point pos,
            Int32Size span, int flag /*= 0*/,
            int border /*= 0*/, IntPtr userData) => default;

        // Adds the given item to the given position.
        public static IntPtr Add3(IntPtr handle, IntPtr item) => default;

        // Adds a spacer to the given position.
        public static IntPtr Add3(IntPtr handle, int width, int height, Int32Point pos,
            Int32Size span, int flag /*= 0*/,
            int border /*= 0*/, IntPtr userData) => default;

        // Look at all items and see if any intersect(or would overlap) the given item.
        public static bool CheckForIntersection(IntPtr handle, IntPtr item,
            IntPtr excludeItem) => default;

        // Look at all items and see if any intersect(or would overlap) the given item.
        public static bool CheckForIntersection2(IntPtr handle, Int32Point pos,
            Int32Size span, IntPtr excludeItem) => default;

        // Find the sizer item for the given window or subsizer, returns NULL if not found.
        public static IntPtr FindItem(IntPtr handle, IntPtr window) => default;

        // Find the sizer item for the given window or subsizer, returns NULL if not found. 
        public static IntPtr FindItem2(IntPtr handle, IntPtr sizer) => default;

        // Get the grid position of the specified item.
        public static Int32Point GetItemPosition(IntPtr handle, IntPtr window) => default;

        // Get the grid position of the specified item.
        public static Int32Point GetItemPosition2(IntPtr handle, IntPtr sizer) => default;

        // Get the grid position of the specified item.
        public static Int32Point GetItemPosition3(IntPtr handle, int index) => default;

        // Get the row/col spanning of the specified item. 
        public static Int32Point GetItemSpan(IntPtr handle, IntPtr window) => default;

        // Get the row/col spanning of the specified item. 
        public static Int32Point GetItemSpan2(IntPtr handle, IntPtr sizer) => default;

        // Get the row/col spanning of the specified item. 
        public static Int32Point GetItemSpan3(IntPtr handle, int index) => default;

        // Set the grid position of the specified item.
        public static bool SetItemPosition(IntPtr handle, IntPtr window, Int32Point pos) => default;

        // Set the grid position of the specified item.
        public static bool SetItemPosition2(IntPtr handle, IntPtr sizer, Int32Point pos) => default;

        // Set the grid position of the specified item.
        public static bool SetItemPosition3(IntPtr handle, int index, Int32Point pos) => default;

        // Set the row/col spanning of the specified item.
        public static bool SetItemSpan(IntPtr handle, IntPtr window, Int32Size span) => default;

        // Set the row/col spanning of the specified item.
        public static bool SetItemSpan2(IntPtr handle, IntPtr sizer, Int32Size span) => default;

        public static bool SetItemSpan3(IntPtr handle, int index, Int32Size span) => default;
    }
}

File: Source/Alternet.UI.Pal/GridBagSizer.cpp
#include "GridBagSizer.h"

namespace Alternet::UI
{
	void* GridBagSizer::CreateGridBagSizer(int vgap, int hgap)
	{
		return new wxGridBagSizer(vgap, hgap);
	}

	Int32Size GridBagSizer::CalcMin(void* handle)
	{
		return ((wxGridBagSizer*)handle)->CalcMin();
	}

	void* GridBagSizer::FindItemAtPoint(void* handle, const Int32Point& pt)
	{
		return ((wxGridBagSizer*)handle)->FindItemAtPoint(pt);
	}

	void* GridBagSizer::FindItemAtPosition(void* handle, const Int32Point& pos)
	{
		wxGBPosition _pos = wxGBPosition(pos.X, pos.Y);
		return ((wxGridBagSizer*)handle)->FindItemAtPosition(_pos);
	}

	void* GridBagSizer::FindItemWithData(void* handle, void* userData)
	{
		return ((wxGridBagSizer*)handle)->FindItemWithData((wxObject*)userData);
	}

	Int32Size GridBagSizer::GetCellSize(void* handle, int row, int col)
	{
		return ((wxGridBagSizer*)handle)->GetCellSize(row, col);
	}

	Int32Size GridBagSizer::GetEmptyCellSize(void* handle)
	{
		return ((wxGridBagSizer*)handle)->GetEmptyCellSize();
	}

	void GridBagSizer::RepositionChildren(void* handle, const Int32Size& minSize)
	{
		((wxGridBagSizer*)handle)->RepositionChildren(minSize);
	}

	void GridBagSizer::SetEmptyCellSize(void* handle, const Int32Size& sz)
	{
		((wxGridBagSizer*)handle)->SetEmptyCellSize(sz);
	}

	void* GridBagSizer::Add(void* handle, void* window, const Int32Point& pos,
		const Int32Size& span, int flag, int border, void* userData)
	{
/*
	wxSizerItem* Add( wxWindow *window,
					  const wxGBPosition& pos,
					  const wxGBSpan& span = wxDefaultSpan,
					  int flag = 0,
					  int border = 0,
					  wxObject* userData = NULL );
*/
		wxGBPosition _pos = wxGBPosition(pos.X, pos.Y);
		wxGBSpan _span = wxGBSpan(span.Width, span.Height);

		return ((wxGridBagSizer*)handle)->Add((wxWindow*) window, _pos,
			_span, flag, border, (wxObject*) userData);
	}

	void* GridBagSizer::Add2(void* handle, void* sizer, const Int32Point& pos,
		const Int32Size& span, int flag, int border, void* userData)
	{
		wxGBPosition _pos = wxGBPosition(pos.X, pos.Y);
		wxGBSpan _span = wxGBSpan(span.Width, span.Height);
		return ((wxGridBagSizer*)handle)->Add((wxSizer*) sizer, _pos,
			_span, flag, border, (wxObject*) userData);
	}

	void* GridBagSizer::Add3(void* handle, void* item)
	{
		return ((wxGridBagSizer*)handle)->Add((wxGBSizerItem*)item);
	}

	void* GridBagSizer::Add4(void* handle, int width, int height,
		const Int32Point& pos, const Int32Size& span, int flag, int border, void* userData)
	{
		wxGBPosition _pos = wxGBPosition(pos.X, pos.Y);
		wxGBSpan _span = wxGBSpan(span.Width, span.Height);
		return ((wxGridBagSizer*)handle)->Add(width, height,
			_pos, _span, flag, border, (wxObject*) userData);
	}

	bool GridBagSizer::CheckForIntersection(void* handle, void* item, void* excludeItem)
	{
/*
	bool CheckForIntersection(wxGBSizerItem* item, wxGBSizerItem* excludeItem = NULL);
	bool CheckForIntersection(const wxGBPosition& pos, const wxGBSpan& span,
	wxGBSizerItem* excludeItem = NULL);
*/

		return ((wxGridBagSizer*)handle)->CheckForIntersection((wxGBSizerItem*)item,
			(wxGBSizerItem*)excludeItem);
	}

	bool GridBagSizer::CheckForIntersection2(void* handle, const Int32Point& pos,
		const Int32Size& span, void* excludeItem)
	{
		wxGBPosition _pos = wxGBPosition(pos.X, pos.Y);
		wxGBSpan _span = wxGBSpan(span.Width, span.Height);
		return ((wxGridBagSizer*)handle)->CheckForIntersection(_pos,
			_span, (wxGBSizerItem*)excludeItem);
	}

	void* GridBagSizer::FindItem(void* handle, void* window)
	{
		return ((wxGridBagSizer*)handle)->FindItem((wxWindow*)window);
	}

	void* GridBagSizer::FindItem2(void* handle, void* sizer)
	{
		return ((wxGridBagSizer*)handle)->FindItem((wxSizer*)sizer);
	}

	Int32Point GridBagSizer::GetItemPosition(void* handle, void* window)
	{
		auto result = ((wxGridBagSizer*)handle)->GetItemPosition((wxWindow*)window);
		return Int32Point(result.GetCol(),result.GetRow());
	}

	Int32Point GridBagSizer::GetItemPosition2(void* handle, void* sizer)
	{
		auto result = ((wxGridBagSizer*)handle)->GetItemPosition((wxSizer*)sizer);
		return Int32Point(result.GetCol(), result.GetRow());
	}

	Int32Point GridBagSizer::GetItemPosition3(void* handle, int index)
	{
		auto result = ((wxGridBagSizer*)handle)->GetItemPosition(index);
		return Int32Point(result.GetCol(), result.GetRow());
	}

	Int32Point GridBagSizer::GetItemSpan(void* handle, void* window)
	{
		auto result = ((wxGridBagSizer*)handle)->GetItemSpan((wxWindow*)window);
		return Int32Point(result.GetColspan(), result.GetRowspan());
	}

	Int32Point GridBagSizer::GetItemSpan2(void* handle, void* sizer)
	{
		auto result = ((wxGridBagSizer*)handle)->GetItemSpan((wxSizer*)sizer);
		return Int32Point(result.GetColspan(), result.GetRowspan());
	}

	Int32Point GridBagSizer::GetItemSpan3(void* handle, int index)
	{
		auto result = ((wxGridBagSizer*)handle)->GetItemSpan(index);
		return Int32Point(result.GetColspan(), result.GetRowspan());
	}

	bool GridBagSizer::SetItemPosition(void* handle, void* window, const Int32Point& pos)
	{
		wxGBPosition _pos = wxGBPosition(pos.X, pos.Y);
		return ((wxGridBagSizer*)handle)->SetItemPosition((wxWindow*) window, _pos);
	}

	bool GridBagSizer::SetItemPosition2(void* handle, void* sizer, const Int32Point& pos)
	{
		wxGBPosition _pos = wxGBPosition(pos.X, pos.Y);
		return ((wxGridBagSizer*)handle)->SetItemPosition((wxSizer*)sizer, _pos);
	}

	bool GridBagSizer::SetItemPosition3(void* handle, int index, const Int32Point& pos)
	{
		wxGBPosition _pos = wxGBPosition(pos.X, pos.Y);
		return ((wxGridBagSizer*)handle)->SetItemPosition(index, _pos);
	}

	bool GridBagSizer::SetItemSpan(void* handle, void* window, const Int32Size& span)
	{
		wxGBSpan _span = wxGBSpan(span.Width, span.Height);
		return ((wxGridBagSizer*)handle)->SetItemSpan((wxWindow*)window, _span);
	}

	bool GridBagSizer::SetItemSpan2(void* handle, void* sizer, const Int32Size& span)
	{
		wxGBSpan _span = wxGBSpan(span.Width, span.Height);
		return ((wxGridBagSizer*)handle)->SetItemSpan((wxSizer*)sizer, _span);
	}

	bool GridBagSizer::SetItemSpan3(void* handle, int index, const Int32Size& span)
	{
		wxGBSpan _span = wxGBSpan(span.Width, span.Height);
		return ((wxGridBagSizer*)handle)->SetItemSpan(index, _span);
	}

	GridBagSizer::GridBagSizer()
	{
	}

	GridBagSizer::~GridBagSizer()
	{
	}
}
